### PR TITLE
record changes in the nodes and update test files

### PR DESCRIPTION
**Pull request by evykassirer**

### CODE REVIEW

File: lib/expression/step-solver/Fraction.js [hunk 1]
```diff
@@ -40,8 +40,10 @@ Fraction.simplifyFraction = function(fraction) {
   if (Negative.isNegative(denominator)) {
     denominator = Negative.negate(denominator);
     numerator = Negative.negate(numerator);
-    const newFraction = NodeCreator.operator('/', [numerator, denominator]);
-    return new NodeStatus(newFraction, MathChangeTypes.SIMPLIFY_FRACTION);
+    let newFraction = NodeCreator.operator('/', [numerator, denominator]);
+    fraction.changeBlock = 0;
+    newFraction.changeBlock = 0;
+    return new NodeStatus(newFraction, MathChangeTypes.SIMPLIFY_FRACTION, fraction);
```
Comment (evykassirer) [inline lib/expression/step-solver/Fraction.js]: I'm considering passing NodeStatus a dictionary now, to make it clear what the parameters are. thoughts?

File: lib/expression/step-solver/NodeStatus.js [hunk 2]
```diff
@@ -1,25 +1,45 @@
...
+
 
 // This represents the current (sub)expression we're simplifying.
 // As we move step by step, a node might be updated. Functions return this
 // status object to pass on the updated node and information on if/how it was
 // changed.
 // NodeStatus(node) creates a NodeStatus object that signals no change
 class NodeStatus {
-  constructor(node, changeType=MathChangeTypes.NO_CHANGE) {
+  constructor(node, changeType=MathChangeTypes.NO_CHANGE,
```
Comment (evykassirer) [inline lib/expression/step-solver/NodeStatus.js]: so many parameters!

File: lib/expression/step-solver/NodeStatus.js [hunk 3]
```diff
@@ -1,25 +1,45 @@
...
+              beforeChangeNode=null, childNodeStatus=null, index=null) {
+    this.node = node;
```
Comment (evykassirer) [inline lib/expression/step-solver/NodeStatus.js]: this is the "after node" - not sure if we want to change the name (in some cases like remove uncessary parens, this is set, but the beforenode won't be)

File: lib/expression/step-solver/NodeStatus.js [hunk 4]
```diff
@@ -1,18 +1,52 @@
...
+
+    if (changeType === MathChangeTypes.NO_CHANGE) {
+      this.hasChanged = false;
+    }
+    else {
+      this.hasChanged = true;
+    }
+
+    this.beforeChangeNode = beforeChangeNode;
+    if (childNodeStatus && childNodeStatus.beforeChangeNode) {
```
Comment (aelnaiem) [inline lib/expression/step-solver/NodeStatus.js]: I don't get it, what's happening here

File: lib/expression/step-solver/ConstantFraction.js [hunk 5]
```diff
@@ -60,10 +62,11 @@ ConstantFraction.simplifyFraction = function(fraction) {
     newFraction = NodeCreator.operator(
       '/', [newNumeratorNode, newDenominatorNode]);
   }
-  fraction.changeBlock = 1;
+  beforeNode.changeBlock = 1;
   newFraction.changeBlock = 1;
 
-  return new NodeStatus(newFraction, MathChangeTypes.SIMPLIFY_FRACTION, fraction);
+  return new NodeStatus(
+    newFraction, MathChangeTypes.SIMPLIFY_FRACTION, beforeNode);
```
Comment (aelnaiem) [inline lib/expression/step-solver/ConstantFraction.js]: Why `newFraction but `beforeNode`? Also I think the order of arguments should be: changetype, newNode, oldNode, so there isn't a weird thing in the middle

**Comment by evykassirer** [comment]
okay it's ready (I think) 

woo

File: lib/expression/step-solver/PolynomialTermNode.js [hunk 6]
```diff
@@ -15,7 +15,7 @@ const NodeType = require('./NodeType');
    (might be null if no exponent)
 */
 class PolynomialTermNode {
-  constructor(node) {
+  constructor(node, onlyImplicit=false) {
```
Comment (aelnaiem) [inline lib/expression/step-solver/PolynomialTermNode.js]: What does `onlyImplicit` doe?

File: lib/expression/step-solver/simplifyExpression.js [hunk 7]
```diff
@@ -18,14 +19,15 @@ const simplifyOperations = require('./simplifyOperations');
 // Given a mathjs expression node, steps through simplifying the expression.
 // Returns the simplified expression node.
 function simplify(node, debug=false) {
+  debug=true;
```
Comment (aelnaiem) [inline lib/expression/step-solver/simplifyExpression.js]: ?

File: lib/expression/step-solver/PolynomialTermNode.js [hunk 8]
```diff
@@ -27,13 +27,19 @@ class PolynomialTermNode {
       }
       // it's '*' ie it has a coefficient
       else if (node.op === '*') {
+        if (onlyImplicit && !node.implicit) {
```
Comment (evykassirer) [inline lib/expression/step-solver/PolynomialTermNode.js]: right here @aelnaiem

File: lib/expression/step-solver/PolynomialTermNode.js [hunk 9]
```diff
@@ -142,9 +148,10 @@ class PolynomialTermNode {
 // e.g. x^2, 2y, z, 3x/5 are all terms. 4, 2+x, 3*7, x-z are all not terms.
 // See the tests for some more thorough examples of exactly what counts and
 // what does not.
-PolynomialTermNode.isPolynomialTerm = function(node, debug=false) {
+PolynomialTermNode.isPolynomialTerm = function(
+    node, debug=false, onlyImplicit=false) {
   try {
-    new PolynomialTermNode(node); // will throw error if node isn't poly term
+    new PolynomialTermNode(node, onlyImplicit); // will throw error if node isn't poly term
```
Comment (evykassirer) [inline lib/expression/step-solver/PolynomialTermNode.js]: and @aelnaiem  here

File: lib/expression/step-solver/flattenOperands.js [hunk 10]
```diff
@@ -163,12 +169,18 @@ function getOperands(node, parentOp) {
     default:
       return [flattenOperands(node)];
   }
+  if (PolynomialTermNode.isPolynomialTerm(node, false, true)) {
```
Comment (evykassirer) [inline lib/expression/step-solver/flattenOperands.js]: and here

File: lib/expression/step-solver/PolynomialTermNode.js (hunk 8 above)
Comment (aelnaiem) [inline lib/expression/step-solver/PolynomialTermNode.js]: No I saw this hahaha, I just can't find where it's set to anything other than false

File: lib/expression/step-solver/flattenOperands.js (hunk 10 above)
Comment (aelnaiem) [inline lib/expression/step-solver/flattenOperands.js]: Cool

**Pull request merged**